archive/imo/imo2019_q2.lean

```diff
@@ -0,0 +1,576 @@
+respectively, we show with an angle chase that `P`, `Q`, `A₂`, `B₂` are concyclic and let `ω` be
+the circle through those points. We then show that `C`, `Q₁`, `A₂`, `A₁` are concyclic, and
+then that `Q₁` lies on `ω`, and similarly that `P₁` lies on `ω`, so the required four points are
+concyclic.
+
+-/
+
+noncomputable theory
+
+open affine affine.simplex euclidean_geometry finite_dimensional
+open_locale affine euclidean_geometry real
+
+local attribute [instance] fact_finite_dimensional_of_finrank_eq_succ
+
+variables (V : Type*) (Pt : Type*) [inner_product_space ℝ V] [metric_space Pt]
+variables [normed_add_torsor V Pt] [hd2 : fact (finrank ℝ V = 2)]
+include hd2
+
+/-- A configuration satisfying the conditions of the problem. We define this structure to avoid
+passing many hypotheses around as we build up information about the configuration; the final
+result for a statement of the problem not using this structure is then deduced from one in terms
+of this structure. -/
+@[nolint has_nonempty_instance]
+structure imo2019q2_cfg :=
+(A B C A₁ B₁ P Q P₁ Q₁ : Pt)
+(affine_independent_ABC : affine_independent ℝ ![A, B, C])
+(A₁_mem_BC : A₁ ∈ affine_segment ℝ B C)
+(B₁_mem_AC : B₁ ∈ affine_segment ℝ A C)
+(P_mem_AA₁ : P ∈ affine_segment ℝ A A₁)
+(Q_mem_BB₁ : Q ∈ affine_segment ℝ B B₁)
+(PQ_parallel_AB : line[ℝ, P, Q] ∥ line[ℝ, A, B])
+-- A hypothesis implicit in the named line.
+(P_ne_Q : P ≠ Q)
+(sbtw_P_B₁_P₁ : sbtw ℝ P B₁ P₁)
+(angle_PP₁C_eq_angle_BAC : ∠ P P₁ C = ∠ B A C)
+-- A hypothesis implicit in the first named angle.
+(C_ne_P₁ : C ≠ P₁)
+(sbtw_Q_A₁_Q₁ : sbtw ℝ Q A₁ Q₁)
+(angle_CQ₁Q_eq_angle_CBA : ∠ C Q₁ Q = ∠ C B A)
+-- A hypothesis implicit in the first named angle.
+(C_ne_Q₁ : C ≠ Q₁)
+
+/-- A default choice of orientation, for lemmas that need to pick one. -/
+def some_orientation : module.oriented ℝ V (fin 2) :=
+⟨basis.orientation (fin_basis_of_finrank_eq _ _ hd2.out)⟩
+
+variables {V Pt}
+
+namespace imo2019q2_cfg
+
+variables (cfg : imo2019q2_cfg V Pt)
+
+/-- The configuration has symmetry, allowing results proved for one point to be applied for
+another (where the informal solution says "similarly"). -/
+def symm : imo2019q2_cfg V Pt :=
+{ A := cfg.B,
+  B := cfg.A,
+  C := cfg.C,
+  A₁ := cfg.B₁,
+  B₁ := cfg.A₁,
+  P := cfg.Q,
+  Q := cfg.P,
+  P₁ := cfg.Q₁,
+  Q₁ := cfg.P₁,
+  affine_independent_ABC := begin
+    rw ←affine_independent_equiv (equiv.swap (0 : fin 3) 1),
+    convert cfg.affine_independent_ABC using 1,
+    ext x,
+    fin_cases x;
+      refl
+  end,
+  A₁_mem_BC := cfg.B₁_mem_AC,
+  B₁_mem_AC := cfg.A₁_mem_BC,
+  P_mem_AA₁ := cfg.Q_mem_BB₁,
+  Q_mem_BB₁ := cfg.P_mem_AA₁,
+  PQ_parallel_AB := set.pair_comm cfg.P cfg.Q ▸ set.pair_comm cfg.A cfg.B ▸ cfg.PQ_parallel_AB,
+  P_ne_Q := cfg.P_ne_Q.symm,
+  sbtw_P_B₁_P₁ := cfg.sbtw_Q_A₁_Q₁,
+  angle_PP₁C_eq_angle_BAC :=
+    angle_comm cfg.C cfg.Q₁ cfg.Q ▸ angle_comm cfg.C cfg.B cfg.A ▸ cfg.angle_CQ₁Q_eq_angle_CBA,
```

Comment:
This and 123 would have been easier with my suggested convention above (not that it matters much)